File: examples/paired_comparison.py
"""Two measurements with a bound predeclared currency-margin comparison."""

from datetime import UTC, datetime
import json
from pathlib import Path

from signal_sdk import ComparisonPlan, ConfirmatoryComparison, FunctionImplementation, MeasurementConfig, measure
from signal_sdk.certificate import compare_measurements
from signal_sdk.domains.payments import generate_book, payments_domain
from signal_sdk.examples import demo_definition, demo_mandate, reconcile


def slightly_different_policy(context):
    result = reconcile(context)
    return f"Completed: {result}"


distribution, episodes = generate_book(24, seed=31, vendors=8)
domain = payments_domain(demo_mandate(episodes))
before = demo_definition(name="policy-before").model_copy(update={"implementation": {"callable": "reconcile", "revision": "1"}})
after = demo_definition(name="policy-after").model_copy(update={"implementation": {"callable": "slightly_different_policy", "revision": "2"}})
plan = ComparisonPlan(declared_at=datetime.now(UTC), comparisons=(
    ConfirmatoryComparison(
        name="wrong-account-loss", reference_id=before.id, candidate_id=after.id,
        metric="loss:wrong_account", margin=1000, maximum_severity=5000,
    ),
))
config = MeasurementConfig(mode="simulation", repetitions=2, seed=19,
                           prepost_plan=plan, bootstrap_samples=300, loss_simulations=300,
                           severity_assumptions={"wrong_account": {"amount": 500, "currency": "USD"}})
pre = measure((FunctionImplementation(before, reconcile, "simulation"),), distribution, episodes, domain=domain, config=config)
post = measure((FunctionImplementation(after, slightly_different_policy, "simulation"),), distribution, episodes,
               domain=domain, config=config)
comparison = compare_measurements(pre, post)
destination = Path("outputs/paired")
destination.mkdir(parents=True, exist_ok=True)
(destination / "pre.json").write_text(pre.model_dump_json(indent=2), encoding="utf-8")
(destination / "post.json").write_text(post.model_dump_json(indent=2), encoding="utf-8")
(destination / "comparison.json").write_text(json.dumps(comparison, indent=2), encoding="utf-8")
print(comparison["paired_differences"][0]["comparisons"][0]["conclusion"])

File: examples/paired-comparison.ts
/**
 * Two measurements with a bound predeclared currency-margin comparison.
 * Run with: npm run build && node dist/examples/paired-comparison.js
 */

import { mkdirSync, writeFileSync } from "node:fs";
import { compareMeasurements } from "../src/certificate.js";
import { generateBook, paymentsEnvironment } from "../src/environments/payments/index.js";
import { demoDefinition, demoMandate, reconcile } from "../src/examples.js";
import { ComparisonPlan, ConfirmatoryComparison, FunctionImplementation, MeasurementConfig, measure } from "../src/index.js";

const slightlyDifferentPolicy: typeof reconcile = (context) => `Completed: ${reconcile(context)}`;

const { distribution, scenarios } = generateBook(24, { seed: 31, vendors: 8 });
const environment = paymentsEnvironment(demoMandate(scenarios));
const before = demoDefinition("policy-before").with({ implementation: { callable: "reconcile", revision: "1" } });
const after = demoDefinition("policy-after").with({ implementation: { callable: "slightlyDifferentPolicy", revision: "2" } });
const plan = new ComparisonPlan({ declaredAt: new Date().toISOString(), comparisons: [
  new ConfirmatoryComparison({ name: "wrong-account-loss", referenceId: before.id, candidateId: after.id, metric: "loss:wrong_account", margin: 1000, maximumSeverity: 5000 }),
] });
const config = new MeasurementConfig({ mode: "simulation", repetitions: 2, seed: 19, prepostPlan: plan, bootstrapSamples: 300, lossSimulations: 300,
  severityAssumptions: { wrong_account: { amount: 500, currency: "USD" } } });
const pre = await measure([new FunctionImplementation(before, reconcile, "simulation")], distribution, scenarios, { environment, config });
const post = await measure([new FunctionImplementation(after, slightlyDifferentPolicy, "simulation")], distribution, scenarios, { environment, config });
const comparison = compareMeasurements(pre, post) as { pairedDifferences: { comparisons: { conclusion: string }[] }[] };
mkdirSync("outputs/paired", { recursive: true });
writeFileSync("outputs/paired/pre.json", JSON.stringify(pre, null, 2));
writeFileSync("outputs/paired/post.json", JSON.stringify(post, null, 2));
writeFileSync("outputs/paired/comparison.json", JSON.stringify(comparison, null, 2));
console.log(comparison.pairedDifferences[0].comparisons[0].conclusion);
